### PR TITLE
split into low-level API and high-level API

### PR DESCRIPTION
**Pull request by derhuerst**
We now have `solve` and `solveUnsafe`.

~~Either this PR or #6 needs to be adapted once the other is merged, so let me know which you will merge first.~~ #6 got merged, merging its code into this PR.

### CODE REVIEW

**Comment by derhuerst** [comment]
The `encode` function now has a different (non-functional) style than its equivalents from before. Let me know if you want the functional style that you used.

File: README.md [hunk 1]
```diff
@@ -1,9 +1,9 @@
-[![Build Status](https://travis-ci.org/dirkschumacher/nodepicosat.svg?branch=master)](https://travis-ci.org/dirkschumacher/nodepicosat)
-[![npm version](https://img.shields.io/npm/v/picosat.svg)](https://www.npmjs.com/package/picosat)
-
 # picosat for node
 
-Node bindings to the [PicoSAT solver release 965](http://fmv.jku.at/picosat/) by Armin Biere. The PicoSAT C code is distributed under a MIT style license and is bundled with this package.
+**Node bindings to the [PicoSAT solver release 965](http://fmv.jku.at/picosat/) by Armin Biere.** The PicoSAT C code is distributed under a MIT style license and is bundled with this package.
+
+[![Build Status](https://travis-ci.org/dirkschumacher/nodepicosat.svg?branch=master)](https://travis-ci.org/dirkschumacher/nodepicosat)
```
Comment (derhuerst) [inline README.md:5]: Moving the badges below the heading and first paragraph makes this the search engine preview readable.

**Comment by dirkschumacher** [comment]
First of all thanks 👍 !

I don't mind adding the string based interface to the package, but I would like to keep the integer based encoding as well. I think this is more natural for people coming from other PicoSAT bindings. Also, when generating formulas automatically, I feel like integers are more efficient. To solve a Sudoku, you already need about 12k clauses.

How about creating a "logical programming" package similiar to this one: https://github.com/sdiehl/picologic ?

**Comment by derhuerst** [comment]
> I don't mind adding the string based interface to the package, but I would like to keep the integer based encoding as well.

The integer- & array-based method or just any integer-based method? **Keep in mind that `solveUnsafe` still takes integer input**, just in the most raw form: two buffers. Only `solve` adds the string-based markup on top of that.

Regarding efficiency: Creating arrays with e.g. 12k integers over and over again isn't very efficient either. This is why chose binary buffers as the low-level input. For example, if you validate many sudokus, you can prevent garbage collection of every single of those 12k arrays by reusing the input buffer.

> How about creating a "logical programming" package similiar to this one: https://github.com/sdiehl/picologic ?

Sounds good, but I'm not sure if it is necessary to make it slightly harder to use this library. The only argument in favour that i can think of is preventing bugs by leveraging existing code.

**Comment by dirkschumacher** [comment]
`solveUnsafe` is fine and a good idea. I just a have a problem with not providing the nested integer array interface anymore. I think it is pretty common among all picosat bindings. IMHO we should keep it.

**Comment by derhuerst** [comment]
Ah, okay. What about accepting the integers as well? For example:

```js
solve([
  // could also be ints only obviously
  ['!A', 1],
  [-1, 'A']
])
```

The only caveats I can think of is that the validation code will be a bit more complex, because it needs to make sure the passed integers and those used for the alphabetical ones won't overlap.

Another solution: A third method.

**Comment by derhuerst** [comment]
Coming from an "outsider" (not familiar with PicoSAT before) perspective, using the alphabetical literals seems is more intuitive to me. At the same time, I can relate to the fact that people familiar with it might want to use the integers that it works with underneath.

**Comment by derhuerst** [comment]
The most recent commits added a `solveWithIntegers` function. It works just like `solve`, but takes integers instead of strings instead.

The commit `foo` was intended to be squashed into `1ca0f88`. 😁

**Comment by dirkschumacher** [comment]
👍 I would like to have it the other way round: `solve` is with integers and `solveWithSymbols??` with letters.

**Comment by derhuerst** [comment]
[Symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) are something else in JavaScript. `solveWithStrings` would be fine IMO.

**Comment by derhuerst** [comment]
For everyone's information: Turns out we had failed completely to do the encoding before (writing `int8`s instead of `int32`s), but the tests didn't fail. fe9c3ef and 594bd1f fix the en- and decoding, 3d26457 adds a tests that reproduces the previous issue.

These commits should be backported to `master`.

**Comment by dirkschumacher** [comment]
Why don't we just merge the whole thing instead of cherry-picking the changes into `master`?

**Comment by derhuerst** [comment]
You can very well do that. Because this PR is a breaking change, you would have to publish as `2.0.0` and `1.x` would be broken forever. :wink:

We could also make this PR non-breaking instead.

**Comment by dirkschumacher** [comment]
What are the breaking changes. Is it just the returned solution object?

**Comment by derhuerst** [comment]
Out of my head, these:

- The result object has a `statusCode` field instead of `status`.
- Code might have depended on the broken behaviour of <=255 variables.

I guess the latter [comes down to a philosophical trade-off](https://softwareengineering.stackexchange.com/questions/368030/is-a-breaking-change-bug-fix-a-semver-patch-or-a-major). Under the circumstances of this repo and this fix, I'd revert the `statusCode` renaming and make it a patch (aka bugfix) release.

**Comment by dirkschumacher** [comment]
Sounds good

**Pull request merged**